src/dune_rules/lock_dir.ml

```diff
@@ -177,11 +177,25 @@ let get_source_path_for_context ctx_name =
       | false -> None
       | true -> Some ctx)
   with
-  | None | Some (Default { lock_dir = None; _ }) -> Memo.return (Some default_source_path)
   | Some (Default { lock_dir = Some lock_dir_selection; _ }) ->
     let+ source_lock_dir = select_lock_dir lock_dir_selection in
     Some source_lock_dir
   | Some (Opam _) -> Memo.return None
+  | None | Some (Default { lock_dir = None; _ }) ->
+    (* CR-someday Alizter: The logic here is hard to follow. Improve the lock
+       directory selection logic so that it is harder to make a mistake. This
+       can be done by enforcing invariants about contexts and their lock
+       directories slightly earlier. *)
+    Memo.return
+    @@
+    (* Context doesn't specify a lock_dir, check if there's a top-level one *)
+    (match workspace.lock_dirs with
+      | [] -> Some default_source_path
+      | [ lock_dir ] -> Some lock_dir.path
+      | _ :: _ :: _ ->
+        (* Multiple lock dirs but context doesn't specify which one to use,
+          fall back to default *)
+        Some default_source_path)
```

Comment:
But what if `default_source_path` is not in `workspace.lock_dirs`?

It *might* be correct; my question is more a behavioral: if the workspace specifies lock dirs, do they exist in addition to the default lock dir or replacing the default lock dir? Personally looking from a user perspective I would assume the latter, but I think it is worth being aware of the choice.